### PR TITLE
Model augmenter

### PR DESCRIPTION
**Pull request by lintusj1**
Create a prior distribution directly from ElfiModel

### CODE REVIEW

File: elfi/methods/bo/acquisition.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ class AcquisitionBase:
                 bounds : tuple of length 'input_dim' of tuples (min, max)
             and methods
                 evaluate(x) : function that returns model (mean, var, std)
-    priors : list of elfi.Priors, optional
+    prior
```
Comment (vuolleko) [inline elfi/methods/bo/acquisition.py]: Type?

File: elfi/methods/bo/utils.py [hunk 2]
```diff
@@ -72,9 +22,9 @@ def minimize(fun, grad, bounds, priors, n_inits=10, maxiter=1000, random_state=N
         Gradient of fun.
     bounds : list of tuples
         Bounds for each parameter.
-    priors : list of elfi.Priors, or list of Nones
-        Used for sampling initialization points. If Nones, sample uniformly. 
-    n_inits : int, optional
+    prior
```
Comment (vuolleko) [inline elfi/methods/bo/utils.py]: Type?

File: elfi/methods/bo/utils.py [hunk 3]
```diff
@@ -3,65 +3,15 @@
 from scipy.optimize import differential_evolution, fmin_l_bfgs_b
 
 
-def approx_second_partial_derivative(fun, x0, dim, h, bounds):
```
Comment (vuolleko) [inline elfi/methods/bo/utils.py]: Why remove this?

File: elfi/methods/utils.py [hunk 4]
```diff
@@ -114,3 +120,79 @@ def _normalize_params(means, weights):
             weights = np.ones(len(means))
         weights = normalize_weights(weights)
         return means, weights
+
+
+# TODO: check that there are no latent variables in parameter parents.
+#       pdfs and gradients wouldn't be correct in those cases as it would require integrating out those latent
+#       variables. This is equivalent to that all stochastic nodes are parameters.
+# TODO: needs some optimization
+class ModelPrior:
```
Comment (vuolleko) [inline elfi/methods/utils.py]: missing documentation

File: elfi/model/augmenter.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    By default this gives the pdfs of the generated model parameters.
+
+    Parameters
+    ----------
+    model : elfi.ElfiModel
+    joint : bool, optional
+        If True (default) return a the joint pdf of the priors
+    log : bool, optional
+        Use logpdf, default False.
```
Comment (vuolleko) [inline elfi/model/augmenter.py]: Using logpdf is numerically more reasonable, so should be default?

File: elfi/methods/methods.py [hunk 6]
```diff
@@ -601,12 +623,17 @@ def plot_state(self, **options):
...
     """Sequential Monte Carlo ABC sampler"""
     def __init__(self, model, discrepancy=None, outputs=None, **kwargs):
-        model, self.discrepancy = self._resolve_model(model, discrepancy)
-        outputs = self._ensure_outputs(outputs, model.parameters + [self.discrepancy])
-        model, added_nodes = self._augment_model(model)
+        model, discrepancy = self._resolve_model(model, discrepancy)
+
+        # Add the prior pdf nodes to the model
+        model = model.copy()
+        pdf = augmenter.add_pdf_nodes(model)[0]
```
Comment (vuolleko) [inline elfi/methods/methods.py]: Using logpdfs would be numerically more reasonable

**Comment by lintusj1** [comment]
Closing temporarily

File: elfi/model/augmenter.py (hunk 5 above)
Comment (lintusj1) [inline elfi/model/augmenter.py]: The name suggests regular pdf, so this is a modifier option.

File: elfi/methods/methods.py (hunk 6 above)
Comment (lintusj1) [inline elfi/methods/methods.py]: Not in this case because it will use sums instead of products.

File: elfi/methods/utils.py (hunk 4 above)
Comment (lintusj1) [inline elfi/methods/utils.py]: Added

File: elfi/loader.py [hunk 7]
```diff
@@ -98,19 +98,19 @@ class RandomStateLoader(Loader):
...
         else:
             raise ValueError("Seed of type {} is not supported".format(seed))
 
         # Jump (or scramble) the state based on batch_index to create parallel separate
         # pseudo random sequences
-        if seed is not False:
-            # TODO: In the future, allow use of https://pypi.python.org/pypi/randomstate ?
+        if seed is not 'global':
+            # TODO: In the future, we could use https://pypi.python.org/pypi/randomstate to enable jumps?
             random_state = np.random.RandomState(get_sub_seed(random_state, batch_index))
```
Comment (lintusj1) [inline elfi/loader.py]: Sorry, could you clarify what you mean?

File: elfi/loader.py [hunk 8]
```diff
@@ -98,19 +98,19 @@ class RandomStateLoader(Loader):
     def load(cls, context, compiled_net, batch_index):
         key = 'output'
         seed = context.seed
-        if seed is False:
+        if seed is 'global':
```
Comment (lintusj1) [inline elfi/loader.py]: I think global is the term for the global numpy PRNG (and in this case it's not really a seed nor used but for debugging and testing).

File: elfi/methods/bo/utils.py (hunk 3 above)
Comment (lintusj1) [inline elfi/methods/bo/utils.py]: It's not used by anything any longer.

File: elfi/methods/bo/utils.py (hunk 2 above)
Comment (lintusj1) [inline elfi/methods/bo/utils.py]: Added

File: elfi/methods/utils.py [hunk 9]
```diff
@@ -116,12 +128,47 @@ def rvs(cls, means, cov=1, weights=None, size=1, random_state=None):
...
+        if means.ndim > 2:
+            raise ValueError('means.ndim = {} but must be at most 2.'.format(means.ndim))
+
         if weights is None:
             weights = np.ones(len(means))
         weights = normalize_weights(weights)
         return means, weights
 
 
+def numgrad(fn, x, h=0.00001):
```
Comment (vuolleko) [inline elfi/methods/utils.py]: TODO: Check robustness

**Pull request merged**